### PR TITLE
Reduce file lookups in node_modules module resolution

### PR DESCRIPTION
**Pull request by andy-ms**
* In `node_modules`, look for a '.d.ts' file before looking for a '.ts' file -- opposite of how it is normally.
* Never look for `node_modules/foo.d.ts`, only `node_modules/foo/index.d.ts`.

_____

***Edit by @DanielRosenwasser***: Addresses issues seen in #16426.

### CODE REVIEW

File: src/compiler/moduleNameResolver.ts [hunk 1]
```diff
@@ -32,10 +32,27 @@ namespace ts {
      * Kinds of file that we are currently looking for.
      * Typically there is one pass with Extensions.TypeScript, then a second pass with Extensions.JavaScript.
      */
-    enum Extensions {
-        TypeScript, /** '.ts', '.tsx', or '.d.ts' */
-        JavaScript, /** '.js' or '.jsx' */
-        DtsOnly /** Only '.d.ts' */
+    const enum Extensions {
```
Comment (yuit) [inline src/compiler/moduleNameResolver.ts]: Name this enum as `Extensions` seems misleading. I was expecting file extensions....though I have no good idea for alternatives 😞

Comment (andy-ms) [inline src/compiler/moduleNameResolver.ts]: How about `ExtensionsToTry`?

**Comment by andy-ms** [comment]
@yuit @aozgaa Could you give this a review?

File: src/compiler/moduleNameResolver.ts (hunk 1 above)
Comment (aozgaa) [inline src/compiler/moduleNameResolver.ts]: Maybe `ExtensionsOfKind`?

File: tests/baselines/reference/moduleResolutionWithExtensions_unexpected2.trace.json [hunk 2]
```diff
@@ -2,30 +2,25 @@
...
-    "File '/node_modules/foo/foo.d.ts' does not exist.",
     "Directory '/node_modules/foo/foo.js' does not exist, skipping all lookups in it.",
+    "File '/node_modules/foo/index.d.ts' does not exist.",
     "File '/node_modules/foo/index.ts' does not exist.",
     "File '/node_modules/foo/index.tsx' does not exist.",
-    "File '/node_modules/foo/index.d.ts' does not exist.",
     "Directory '/node_modules/@types' does not exist, skipping all lookups in it.",
     "Loading module 'foo' from 'node_modules' folder, target file type 'JavaScript'.",
-    "File '/node_modules/foo.js' does not exist.",
-    "File '/node_modules/foo.jsx' does not exist.",
```
Comment (sheetalkamat) [inline tests/baselines/reference/moduleResolutionWithExtensions_unexpected2.trace.json]: removal of this path lookup seems incorrect.
Refer https://nodejs.org/api/modules.html

> 
> LOAD_NODE_MODULES(X, START)
> 1. let DIRS=NODE_MODULES_PATHS(START)
> 2. for each DIR in DIRS:
>    **a. LOAD_AS_FILE(DIR/X)**
>    b. LOAD_AS_DIRECTORY(DIR/X)

It looks like node does look for node_modules/foo.js

Comment (andy-ms) [inline tests/baselines/reference/moduleResolutionWithExtensions_unexpected2.trace.json]: That's true, but as far as I know, every package manager fills `node_modules` with only directories, and searching for a file there is costing us a lot of `fileExists` calls (as you can see from the diff).

Comment (aozgaa) [inline tests/baselines/reference/moduleResolutionWithExtensions_unexpected2.trace.json]: Pinging @DanielRosenwasser , as he may have some more insight into whether we want to do this.

File: src/compiler/moduleNameResolver.ts [hunk 3]
```diff
@@ -32,10 +32,27 @@ namespace ts {
      * Kinds of file that we are currently looking for.
      * Typically there is one pass with Extensions.TypeScript, then a second pass with Extensions.JavaScript.
```
Comment (aozgaa) [inline src/compiler/moduleNameResolver.ts]: Add a comment explaining the significance of the enum's members' values.

File: src/compiler/moduleNameResolver.ts [hunk 4]
```diff
@@ -32,10 +32,27 @@ namespace ts {
...
-    enum Extensions {
-        TypeScript, /** '.ts', '.tsx', or '.d.ts' */
-        JavaScript, /** '.js' or '.jsx' */
-        DtsOnly /** Only '.d.ts' */
+    const enum Extensions {
+        /**
+         * '.ts', '.tsx', or '.d.ts'
+         * This is for regular imports of TypeScript code.
+         */
+        TypeScript = "TypeScript",
```
Comment (aozgaa) [inline src/compiler/moduleNameResolver.ts]: Would adding `('.ts' preferred)` make sense here? May clarify debugging for people looking at trace-reoslution output.

File: src/compiler/moduleNameResolver.ts (hunk 3 above)
Comment (andy-ms) [inline src/compiler/moduleNameResolver.ts]: This is pretty heavily commented already, what should I add?

File: tests/baselines/reference/moduleResolutionWithExtensions_unexpected2.trace.json (hunk 2 above)
Comment (rbuckton) [inline tests/baselines/reference/moduleResolutionWithExtensions_unexpected2.trace.json]: Its unlikely we will break anyone with this change, but I'd like to see what kind of performance gain we get out of removing these extraneous lookups.

Comment (andy-ms) [inline tests/baselines/reference/moduleResolutionWithExtensions_unexpected2.trace.json]: I ran the perf tests and there is basically no change between this branch and `master` -- a 0.5% improvement in parse time.

**Comment by mhegazy** [comment]
@andy-ms please refresh this PR.

**Pull request closed without merging**